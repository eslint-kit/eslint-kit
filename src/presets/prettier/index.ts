import { Config } from 'prettier'
import { publicPresetNames } from '../names'
import { createPreset } from '../shared'

export const prettier = createPreset<Config>({
  name: publicPresetNames.prettier,
  compile: ({ options }) => ({
    plugins: ['prettier'],
    rules: {
      'prettier/prettier': options ? ['warn', options] : 'warn',
    },
  }),
})
