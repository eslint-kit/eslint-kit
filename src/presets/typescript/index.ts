import path from 'path'
import { Linter } from '@typescript-eslint/utils/dist/ts-eslint'
import { EXTENSIONS, importExtensionsRule } from '../../shared'
import { createPreset } from '../shared'

export interface Options {
  root?: string
  tsconfig?: string
}

export const typescriptRules: Linter.RulesRecord = {
  '@typescript-eslint/adjacent-overload-signatures': 'warn',
  '@typescript-eslint/ban-ts-comment': 'error',
  '@typescript-eslint/ban-types': 'error',
  '@typescript-eslint/naming-convention': [
    'warn',
    {
      selector: 'variable',
      format: ['camelCase', 'UPPER_CASE', 'PascalCase'],
      leadingUnderscore: 'allow',
    },
    {
      selector: 'parameter',
      format: ['camelCase'],
      leadingUnderscore: 'allow',
    },
    {
      selector: 'typeLike',
      format: ['PascalCase'],
    },
  ],
  '@typescript-eslint/consistent-type-assertions': 'warn',
  '@typescript-eslint/consistent-type-definitions': ['warn', 'interface'],
  '@typescript-eslint/no-array-constructor': 'warn',
  '@typescript-eslint/no-empty-interface': 'warn',
  '@typescript-eslint/no-explicit-any': 'warn',
  '@typescript-eslint/no-extra-non-null-assertion': 'warn',
  '@typescript-eslint/no-for-in-array': 'warn',
  '@typescript-eslint/no-inferrable-types': 'error',
  '@typescript-eslint/no-misused-new': 'error',
  '@typescript-eslint/no-namespace': 'error',
  '@typescript-eslint/no-non-null-assertion': 'warn',
  '@typescript-eslint/no-unused-vars': [
    'warn',
    {
      ignoreRestSiblings: true,
      varsIgnorePattern: '^_',
      argsIgnorePattern: '^_',
    },
  ],
  '@typescript-eslint/no-unnecessary-condition': 'warn',
  '@typescript-eslint/no-unused-expressions': 'warn',
  '@typescript-eslint/no-useless-constructor': 'warn',
  '@typescript-eslint/prefer-for-of': 'warn',
  '@typescript-eslint/prefer-includes': 'warn',
  '@typescript-eslint/prefer-namespace-keyword': 'error',
  '@typescript-eslint/prefer-optional-chain': 'warn',
  '@typescript-eslint/prefer-string-starts-ends-with': 'warn',
  '@typescript-eslint/require-array-sort-compare': 'error',
  '@typescript-eslint/return-await': ['error', 'in-try-catch'],
  '@typescript-eslint/triple-slash-reference': 'error',

  // Specific rules that are enabled using @typescript-eslint, but have analogues in common eslint
  'camelcase': 'off',
  'no-array-constructor': 'off',
  'no-unused-vars': 'off',
  'no-unused-expressions': 'off',
  'no-useless-constructor': 'off',
  'no-return-await': 'off',

  // Checked by Typescript - ts(2378)
  'getter-return': 'off',
  // Checked by Typescript - ts(2300)
  'no-dupe-args': 'off',
  // Checked by Typescript - ts(1117)
  'no-dupe-keys': 'off',
  // Checked by Typescript - ts(7027)
  'no-unreachable': 'off',
  // Checked by Typescript - ts(2367)
  'valid-typeof': 'off',
  // Checked by Typescript - ts(2588)
  'no-const-assign': 'off',
  // Checked by Typescript - ts(2588)
  'no-new-symbol': 'off',
  // Checked by Typescript - ts(2376)
  'no-this-before-super': 'off',
  // This is checked by Typescript using the option `strictNullChecks`.
  'no-undef': 'off',
  // This is already checked by Typescript.
  'no-dupe-class-members': 'off',
  // This is already checked by Typescript.
  'no-redeclare': 'off',

  // Checked by Typescript
  'import/no-extraneous-dependencies': 'off',
}

export const typescript = createPreset<'typescript', Options | void>({
  name: 'typescript',
  updateMeta: ({ meta, options }) => {
    meta.extensions = meta.extensions.concat(EXTENSIONS.TS)
    meta.typescript.used = true
    meta.typescript.root = path.resolve(meta.root, options?.root ?? './')
    meta.typescript.tsconfig = options?.tsconfig ?? 'tsconfig.json'
  },
  compile: ({ meta }) => ({
    plugins: ['import', '@typescript-eslint'],
    parser: '@typescript-eslint/parser',
    parserOptions: {
      project: meta.typescript.tsconfig,
      createDefaultProgram: true,
      tsconfigRootDir: meta.typescript.root,
    },
    settings: {
      'import/extensions': meta.extensions,
      'import/resolver': {
        node: {
          extensions: meta.extensions,
        },
      },
      'import/ignore': ['\\.(coffee|scss|css|less|hbs|svg|json)$'],
      'import/external-module-folders': ['node_modules', 'node_modules/@types'],
      'import/parsers': {
        '@typescript-eslint/parser': EXTENSIONS.TS,
      },
    },
    rules: {
      'import/extensions': [
        importExtensionsRule[0],
        importExtensionsRule[1],
        meta.extensions.reduce<{ [key: string]: 'never' }>((acc, extension) => {
          const extensionShort = extension.slice(1)
          acc[extensionShort] = 'never'
          return acc
        }, {}),
      ],
    },
    overrides: [
      {
        files: ['*.ts', '*.tsx'],
        rules: typescriptRules,
      },
    ],
  }),
})
