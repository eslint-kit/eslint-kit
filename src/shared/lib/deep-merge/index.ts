/* eslint-disable @typescript-eslint/no-explicit-any */
import { Merge, PathsOf } from './types'

function isObject(value: any): value is Record<string, any> {
  return typeof value === 'object' && value !== null
}

export enum Strategy {
  Shallow,
  Deep,
  Override,
}

type AllPaths<A, B> = PathsOf<A> | PathsOf<B> | null

function shallowMerge<A, B>(a: A, b: B): Merge<A, B> {
  let merged: any

  if (Array.isArray(a) && Array.isArray(b)) {
    merged = [...a, ...b]
  } else if (isObject(a) && isObject(b)) {
    merged = { ...a, ...b }
  } else if (typeof b !== 'undefined') {
    merged = b
  } else {
    merged = a
  }

  return merged as Merge<A, B>
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function deepMerge<A, B>(
  a: A,
  b: B,
  strategy: (path: AllPaths<A, B>) => Strategy = () => Strategy.Deep,
  basePath: AllPaths<A, B> = null,
): Merge<A, B> {
  if (strategy(basePath) === Strategy.Shallow) {
    return shallowMerge(a, b)
  }

  if (strategy(basePath) === Strategy.Override) {
    return b as Merge<A, B>
  }

  if (Array.isArray(a) && Array.isArray(b)) {
    return shallowMerge(a, b)
  }

  if (isObject(a) && isObject(b)) {
    const keys = [...Object.keys(a), ...Object.keys(b)]
    const final: any = {}

    for (const key of keys) {
      const path = basePath ? [basePath, key].join('.') : key
      final[key] = deepMerge(
        (a as any)[key],
        (b as any)[key],
        strategy,
        path as AllPaths<A, B>,
      )
    }

    return final
  }

  return shallowMerge(a, b)
}
