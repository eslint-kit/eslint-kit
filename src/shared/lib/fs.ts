import fs from 'fs'
import path from 'path'

export function readJson<T>(root: string, file: string): T | null {
  const fullpath = path.resolve(root, file)
  if (!fs.existsSync(fullpath)) return null
  const data = fs.readFileSync(fullpath, 'utf-8')
  try {
    return JSON.parse(data)
  } catch {
    return null
  }
}
