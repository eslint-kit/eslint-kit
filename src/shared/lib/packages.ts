import { parse } from 'semver'
import { PackageJson } from '../types'

function mergeDependencies(packageJson: PackageJson) {
  const { dependencies = {}, devDependencies = {} } = packageJson
  return { ...dependencies, ...devDependencies }
}

export function getPackageSemver(name: string, packageJson: PackageJson) {
  const allDependencies = mergeDependencies(packageJson)
  const version = allDependencies[name]
  if (!version) return null
  const parsed = parse(version)
  if (!parsed) return null
  return parsed
}
