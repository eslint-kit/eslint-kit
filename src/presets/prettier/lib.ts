import { Linter } from 'eslint'
import { Options } from 'prettier'
import { Meta } from '../shared'

export interface Input {
  options?: Options
  meta: Meta
}

export function createPrettierRule({ meta, options }: Input) {
  const plugins: string[] = []

  if (meta.imports.layout) {
    plugins.push('prettier-plugin-layout')
  }

  return [
    'warn',
    {
      plugins,
      ...options,
    },
  ] as Linter.RuleEntry
}
