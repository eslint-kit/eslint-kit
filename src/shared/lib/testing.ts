import fs from 'fs'
import path from 'path'
import { ESLint } from 'eslint'
import { Preset, PresetName, PRIORITY, compilePresets } from '../../presets'
import { base } from '../../presets/base'

export interface Options {
  presets: Preset[]
  priority?: PresetName[]
  dirname: string
  files: string[]
  extension: 'js' | 'jsx' | 'ts' | 'tsx' | 'svelte' | 'vue'
}

export async function testConfig({
  presets,
  priority = PRIORITY,
  dirname,
  files,
  extension,
}: Options) {
  const root = path.resolve(dirname, './tests')

  const cli = new ESLint({
    baseConfig: compilePresets([base({ root }), ...presets], priority),
    useEslintrc: false,
    cwd: root,
    ignore: false,
  })

  for (const file of files) {
    const filePath = path.resolve(root, `./${file}.${extension}`)
    const code = fs.readFileSync(filePath).toString()
    const result = await cli.lintText(code, { filePath })
    expect(result).toMatchSnapshot(`${file}`)
  }
}
