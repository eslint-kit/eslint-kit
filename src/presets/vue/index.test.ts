import { testConfig } from '../../shared/lib/testing'
import { prettier } from '../prettier'
import { typescript } from '../typescript'
import { vue } from './index'

describe('[Presets] Vue', () => {
  it('should match snapshots', async () => {
    await testConfig({
      presets: [vue()],
      dirname: __dirname,
      files: ['basic-clear', 'basic-warn', 'basic-error'],
      extension: 'js',
    })
  })

  it('should lint .vue files', async () => {
    await testConfig({
      presets: [vue()],
      dirname: __dirname,
      files: ['vue-clear', 'vue-warn', 'vue-error'],
      extension: 'vue',
    })
  })

  it('should work with typescript', async () => {
    await testConfig({
      presets: [typescript(), vue()],
      dirname: __dirname,
      files: ['basic-clear', 'basic-warn', 'basic-error'],
      extension: 'ts',
    })
  })

  it('should lint typescript in .vue files', async () => {
    await testConfig({
      presets: [typescript(), vue()],
      dirname: __dirname,
      files: ['ts-vue-clear', 'ts-vue-warn', 'ts-vue-error'],
      extension: 'vue',
    })
  })

  it('should not conflict with prettier', async () => {
    await testConfig({
      presets: [
        prettier({
          semi: false,
          singleQuote: true,
          tabWidth: 2,
          quoteProps: 'consistent',
          endOfLine: 'lf',
        }),
        vue(),
      ],
      dirname: __dirname,
      files: ['vue-warn'],
      extension: 'vue',
    })
  })
})
