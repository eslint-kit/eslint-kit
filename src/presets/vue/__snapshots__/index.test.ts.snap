// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Presets] Vue should lint .vue files: vue-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/vue-clear.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 2,
    "messages": Array [
      Object {
        "column": 30,
        "endColumn": 30,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            332,
            332,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 51,
        "endColumn": 51,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            353,
            353,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
const count = ref(0)

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;

exports[`[Presets] Vue should lint .vue files: vue-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/vue-error.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 2,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 28,
        "endLine": 5,
        "line": 5,
        "message": "\`<script setup>\` cannot contain ES module exports.",
        "messageId": "forbidden",
        "nodeType": "ExportNamedDeclaration",
        "ruleId": "vue/no-export-in-script-setup",
        "severity": 2,
      },
      Object {
        "column": 30,
        "endColumn": 30,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            339,
            339,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 51,
        "endColumn": 51,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            360,
            360,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
export const count = ref(0)

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;

exports[`[Presets] Vue should lint .vue files: vue-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/vue-warn.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 4,
    "messages": Array [
      Object {
        "column": 22,
        "endColumn": 32,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            313,
            334,
          ],
          "text": "id=\\"hello\\" ref=\\"meow\\"",
        },
        "line": 19,
        "message": "Attribute \\"id\\" should go before \\"ref\\".",
        "messageId": "expectedOrder",
        "nodeType": "VAttribute",
        "ruleId": "vue/attributes-order",
        "severity": 1,
      },
      Object {
        "column": 22,
        "endColumn": 32,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            323,
            324,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "'id' should be on a new line.",
        "messageId": "shouldBeOnNewLine",
        "nodeType": "VAttribute",
        "ruleId": "vue/max-attributes-per-line",
        "severity": 1,
      },
      Object {
        "column": 33,
        "endColumn": 51,
        "endLine": 19,
        "line": 19,
        "message": "'@click' should be on a new line.",
        "messageId": "shouldBeOnNewLine",
        "nodeType": "VAttribute",
        "ruleId": "vue/max-attributes-per-line",
        "severity": 1,
      },
      Object {
        "column": 52,
        "endColumn": 52,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            354,
            354,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 73,
        "endColumn": 73,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            375,
            375,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
const count = ref(0)

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button ref=\\"meow\\" id=\\"hello\\" @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 5,
  },
]
`;

exports[`[Presets] Vue should lint typescript in .vue files: ts-vue-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/ts-vue-clear.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Vue should lint typescript in .vue files: ts-vue-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/ts-vue-error.vue",
    "fixableErrorCount": 1,
    "fixableWarningCount": 2,
    "messages": Array [
      Object {
        "column": 14,
        "endColumn": 20,
        "endLine": 5,
        "fix": Object {
          "range": Array [
            84,
            90,
          ],
          "text": "number",
        },
        "line": 5,
        "message": "Don't use \`Number\` as a type. Use number instead",
        "messageId": "bannedTypeMessage",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/ban-types",
        "severity": 2,
      },
      Object {
        "column": 30,
        "endColumn": 30,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            340,
            340,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 51,
        "endColumn": 51,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            361,
            361,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
const count: Number = ref(0)

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;

exports[`[Presets] Vue should lint typescript in .vue files: ts-vue-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/ts-vue-warn.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 2,
    "messages": Array [
      Object {
        "column": 25,
        "endColumn": 28,
        "endLine": 5,
        "line": 5,
        "message": "Unexpected any. Specify a different type.",
        "messageId": "unexpectedAny",
        "nodeType": "TSAnyKeyword",
        "ruleId": "@typescript-eslint/no-explicit-any",
        "severity": 1,
        "suggestions": Array [
          Object {
            "desc": "Use \`unknown\` instead, this will force you to explicitly, and safely assert the type is correct.",
            "fix": Object {
              "range": Array [
                95,
                98,
              ],
              "text": "unknown",
            },
            "messageId": "suggestUnknown",
          },
          Object {
            "desc": "Use \`never\` instead, this is useful when instantiating generic type parameters that you don't need to know the type of.",
            "fix": Object {
              "range": Array [
                95,
                98,
              ],
              "text": "never",
            },
            "messageId": "suggestNever",
          },
        ],
      },
      Object {
        "column": 30,
        "endColumn": 30,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            339,
            339,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 51,
        "endColumn": 51,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            360,
            360,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
const count = ref(0) as any

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 3,
  },
]
`;

exports[`[Presets] Vue should match snapshots: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Vue should match snapshots: basic-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-error.js",
    "fixableErrorCount": 1,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 12,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            0,
            3,
          ],
          "text": "let",
        },
        "line": 1,
        "message": "Unexpected var, use let or const instead.",
        "messageId": "unexpectedVar",
        "nodeType": "VariableDeclaration",
        "ruleId": "no-var",
        "severity": 2,
      },
    ],
    "source": "var one = 1
export { one }
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Vue should match snapshots: basic-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-warn.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 7,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "'one' is assigned a value but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "const one = 1
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 1,
  },
]
`;

exports[`[Presets] Vue should not conflict with prettier: vue-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/vue-warn.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 6,
    "messages": Array [
      Object {
        "column": 22,
        "endColumn": 32,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            313,
            334,
          ],
          "text": "id=\\"hello\\" ref=\\"meow\\"",
        },
        "line": 19,
        "message": "Attribute \\"id\\" should go before \\"ref\\".",
        "messageId": "expectedOrder",
        "nodeType": "VAttribute",
        "ruleId": "vue/attributes-order",
        "severity": 1,
      },
      Object {
        "column": 22,
        "endColumn": 32,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            323,
            324,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "'id' should be on a new line.",
        "messageId": "shouldBeOnNewLine",
        "nodeType": "VAttribute",
        "ruleId": "vue/max-attributes-per-line",
        "severity": 1,
      },
      Object {
        "column": 33,
        "endColumn": 51,
        "endLine": 19,
        "line": 19,
        "message": "'@click' should be on a new line.",
        "messageId": "shouldBeOnNewLine",
        "nodeType": "VAttribute",
        "ruleId": "vue/max-attributes-per-line",
        "severity": 1,
      },
      Object {
        "column": 52,
        "endColumn": 73,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            354,
            375,
          ],
          "text": "
    Count is: {{ count }}
  ",
        },
        "line": 19,
        "message": "Replace \`Count·is:·{{·count·}}\` with \`⏎····Count·is:·{{·count·}}⏎··\`",
        "messageId": "replace",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
      Object {
        "column": 52,
        "endColumn": 52,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            354,
            354,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break after opening tag (\`<button>\`), but no line breaks found.",
        "messageId": "unexpectedAfterClosingBracket",
        "nodeType": "HTMLTagClose",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 73,
        "endColumn": 73,
        "endLine": 19,
        "fix": Object {
          "range": Array [
            375,
            375,
          ],
          "text": "
",
        },
        "line": 19,
        "message": "Expected 1 line break before closing tag (\`</button>\`), but no line breaks found.",
        "messageId": "unexpectedBeforeOpeningBracket",
        "nodeType": "HTMLEndTagOpen",
        "ruleId": "vue/singleline-html-element-content-newline",
        "severity": 1,
      },
      Object {
        "column": 12,
        "endColumn": 12,
        "endLine": 20,
        "fix": Object {
          "range": Array [
            396,
            396,
          ],
          "text": "
",
        },
        "line": 20,
        "message": "Insert \`⏎\`",
        "messageId": "insert",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
    ],
    "source": "<script setup>
import { ref, onMounted } from 'vue'

// reactive state
const count = ref(0)

// functions that mutate state and trigger updates
function increment() {
  count.value++
}

// lifecycle hooks
onMounted(() => {
  console.info(\`The initial count is \${count.value}.\`)
})
</script>

<template>
  <button ref=\\"meow\\" id=\\"hello\\" @click=\\"increment\\">Count is: {{ count }}</button>
</template>",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 7,
  },
]
`;

exports[`[Presets] Vue should work with typescript: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-clear.ts",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Vue should work with typescript: basic-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-error.ts",
    "fixableErrorCount": 1,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 12,
        "endColumn": 18,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            11,
            17,
          ],
          "text": "number",
        },
        "line": 1,
        "message": "Don't use \`Number\` as a type. Use number instead",
        "messageId": "bannedTypeMessage",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/ban-types",
        "severity": 2,
      },
    ],
    "source": "const one: Number = 1
export { one }
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Vue should work with typescript: basic-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/vue/tests/basic-warn.ts",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 6,
        "endColumn": 10,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            0,
            12,
          ],
          "text": "interface Test ",
        },
        "line": 1,
        "message": "Use an \`interface\` instead of a \`type\`.",
        "messageId": "interfaceOverType",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/consistent-type-definitions",
        "severity": 1,
      },
      Object {
        "column": 6,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "'Test' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": null,
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "type Test = {
  one: number
}
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;
